gh-89727: Fix os.fwalk RecursionError on deep trees

jonburdo opened the pull request:
Use a stack to implement os.fwalk iteratively instead of recursively to avoid hitting recursion limits on deeply nested trees.

Similar to how this is done for `os.walk` in https://github.com/python/cpython/pull/99803

<!-- gh-issue-number: gh-89727 -->
* Issue: gh-89727
<!-- /gh-issue-number -->

jonburdo committed:
fix os.walk RecursionError on deep trees

jonburdo committed:
use _WalkAction enum

jonburdo committed:
Merge branch 'main' into iterative-os-fwalk

AlexWaygood commented on Lib/os.py:
```diff
@@ -427,6 +427,13 @@ def _walk(top, topdown, onerror, followlinks):
+    class _WalkAction:
+        YIELD = 1
+        CLOSE = 2
+        WALK = 3
```
This might lead to performance regressions, since `enum` is known to have slow import times, and `os` is quite a low-level module that's imported by a lot of other modules.

In this case, it looks like this probably doesn't _need_ to be an enum. You could maybe achieve the same thing just like this:

```suggestion
    class _WalkAction:
        YIELD = object()
        CLOSE = object()
        WALK = object()
```

jonburdo replied:
Yeah, good call - thanks!

jonburdo committed:
gh-89727: Fix os.walk RecursionError on deep trees (#99803)

Use a stack to implement os.walk iteratively instead of recursively to
avoid hitting recursion limits on deeply nested trees.

slateny committed:
gh-69929: re docs: Add more specific definition of \w (#92015)

Co-authored-by: Jelle Zijlstra <jelle.zijlstra@gmail.com>

graingert committed:
gh-89051: Add ssl.OP_LEGACY_SERVER_CONNECT (#93927)

Co-authored-by: blurb-it[bot] <43283697+blurb-it[bot]@users.noreply.github.com>
Co-authored-by: Christian Heimes <christian@python.org>
Co-authored-by: Hugo van Kemenade <hugovk@users.noreply.github.com>
Fixes https://github.com/python/cpython/issues/89051

tbwolfe committed:
gh-88211: Change lower-case and upper-case to match recommendations in imaplib docs (#99625)

rkojedzinszky committed:
gh-100348: Fix ref cycle in `asyncio._SelectorSocketTransport` with `_read_ready_cb` (#100349)

fnesveda committed:
gh-99925: Fix inconsistency in `json.dumps()` error messages (GH-99926)

pablogsal committed:
Clarify that every thread has its own default context in contextvars (#99246)

jonburdo committed:
fix formatting

jonburdo committed:
remove enum from os

jonburdo committed:
add blurb

jonburdo committed:
run test_walk_above_recursion_limit for fwalk

jonburdo committed:
set stack outside try-except in fwalk

jonburdo committed:
fix comments

jonburdo committed:
Merge branch 'main' into iterative-os-fwalk

AlexWaygood commented on Lib/os.py:
```diff
@@ -485,58 +489,81 @@ def fwalk(top=".", topdown=True, onerror=None, *, follow_symlinks=False, dir_fd=
-                        nondirs.append(name)
-                except OSError:
-                    pass
+                    raise ValueError(f"invalid walk action: {action}")
```
A `ValueError` should generally be used for bad user input of some kind. This doesn't look like that: we should only get here (if I understand correctly) if there's been some kind of programming error on the part of the people writing this function. An assertion might be more idiomatic here:

```suggestion
                    assert False, f"invalid walk action: {action}"
```

jonburdo replied:
Sounds good. Yeah, this theoretically should never happen. Maybe just `raise AssertionError` would be more direct than `assert False`? Seems like that's the more common way to do it as well when the condition has already been proven False

AlexWaygood replied:
Some people care very passionately about the difference between `assert False` and `raise AssertionError`, but I am not one of them -- either is fine by me :)

jonburdo committed:
change ValueError to AssertionError

jonburdo committed:
Merge branch 'main' into iterative-os-fwalk

jonburdo committed:
add more reliable file descriptor closing logic in os.fwalk
Make sure file descriptors get closed if an exception occurs right after the file descriptor was popped from the stack. Also catch exceptions thrown by calling os.close on an already-closed file-descriptor when doing file descriptor cleanup.

jonburdo committed:
use a separate fd_stack for simpler cleanup and error handling

jonburdo committed:
Merge branch 'main' into iterative-os-fwalk

jonburdo committed:
run test_walk_above_recursion_limit with fwalk tests

barneygale commented on Lib/os.py:
```diff
@@ -485,58 +489,89 @@ def fwalk(top=".", topdown=True, onerror=None, *, follow_symlinks=False, dir_fd=
-        # #13734.
+        # bpo-13734 (gh-57943).
+        stack = [(_WalkAction.WALK, (topfd, toppath))]
+        fd_stack = []
```
Could you store the fd to close as the second argument of stack tuples, like `(_WalkAction.CLOSE, fd)`, rather than using a separate stack? It would mean you'd need to traverse `stack` in your `finally:` block, but that's probably OK?

jonburdo replied:
We could with a couple of caveats. I actually originally did that, but then realized that the fd wouldn't get closed if an exception is raised immediately after it's popped from the stack (i.e. it won't get handled in the `finally`). It also means iterating through a longer list and checking the action type for `_WalkAction.CLOSE` in the `finally` - like you said probably not a big deal, but having `fd_stack` does make this cleanup step simpler.

See the commit where I changed this for the difference: https://github.com/python/cpython/pull/100347/commits/f0f9330f4c2a6bbb24957002f1442ca135022c37
Note, even before this commit, to make sure something like a `KeyboardException` doesn't leave the fd open, we probably need change the code to first check the top of the stack for an fd, close it, and then pop the item:
```python
action, value = stack[-1]
...
stack.pop()
```
Maybe this isn't critical, but I figured a second list was simple enough. Lmk if you still think it's worth getting rid of `fd_stack`.

jonburdo replied:
I think I was maybe looking at this wrong. By popping after calling `close`, we risk calling `close` twice on the same fd. Anyway, I switched back to using one `stack`. That simplifies a little and optimizes for the case where we don't get an exception. Iterating over the entire stack in the `except` should be an acceptable cost I think

jonburdo committed:
Merge branch 'main' into iterative-os-fwalk

jonburdo committed:
make sure we don't close the same fd twice

If an exception occurred between `close(fd_stack[-1])` and `fd_stack.pop()`, then we would close the fd a second time in the except clause. It would be better to risk leaving the fd open, as the previous implementation of fwalk could

jonburdo committed:
revert to using a single stack instead of a separate ffd stack

jonburdo committed:
remove unused variable

barneygale reviewed (approved):
LGTM, nice work!

barneygale commented on Lib/os.py:
```diff
@@ -485,58 +489,85 @@ def fwalk(top=".", topdown=True, onerror=None, *, follow_symlinks=False, dir_fd=
+                        close(value)
+                    except OSError:
+                        pass
+            raise
```
Could this be a `finally:` block?

jonburdo replied:
Sure, would you consider that more readable or explicit? I just didn't see any benefit, since this is only needed if an exception occurs. The only functional difference would be that we'd iterate over an empty list if no exceptions occur right?

barneygale replied:
I think I'd consider it more explicit, though feel free to wait for another opinion if you'd like.

AlexWaygood replied:
I'd also prefer `finally` here. Not sure if it actually delivers any functional benefit here (can't remember the exact semantics off the top of my head...), but it does feel at least more readable to me.

jonburdo replied:
Makes sense, thanks for clarifying. I'll change it to `finally`. Functionally, I think `finally` is slightly worse but it's negligible.

In this case, changing to `finally` is functionally equivalent to adding an `else` clause with the same for-loop:
```python
        try:
            ...
        except:
            for action, value in reversed(stack): ...
            raise
        else:
            for action, value in reversed(stack): ...
```
If no exception occurs in the `try`, then `stack` will be empty and all of the fds will have been closed already. So changing to `finally` means we're adding a step where we simply iterate over an empty list - unnecessary, but costs almost nothing.

To me, a bare `except` indicates a cleanup step, but `finally` is even more explicit, so it probably makes sense here.

jonburdo committed:
get rid of unnecessary os._fwalk

jonburdo committed:
change except clause to finally clause in os.fwalk

AlexWaygood committed:
Merge branch 'main' into iterative-os-fwalk

erlend-aasland committed:
Merge branch 'main' into iterative-os-fwalk
